orchard/orchard_runner.go

```diff
@@ -8,26 +8,100 @@ package orchard
 import (
 	"encoding/json"
 	"fmt"
+	"log"
+	"os"
 	"os/exec"
+	"strings"
+
+	"github.com/google/uuid"
+	"mce.salesforce.com/sprinkler/common"
 )
 
+const baseDir string = "/sprinkler"
+
 type OrchardRunner interface {
-	Generate(artifact string, command string) string
+	Generate(artifact string, command string) (string, error)
 }
 
 type OrchardStdoutRunner struct{}
 
-func (self OrchardStdoutRunner) Generate(artifact string, command string) (string, error) {
-	cmds, err := parseCommandLine(command)
+func (r OrchardStdoutRunner) Generate(artifact string, command string) (string, error) {
+	if artifact == "" {
+		return processCmd(command, baseDir)
+	}
+
+	if !strings.HasPrefix(artifact, "s3://") {
+		return "", fmt.Errorf("artifact %v is not supported\n", artifact)
+	}
+
+	return s3ArtifactGenerate(artifact, command)
+}
+
+func s3ArtifactGenerate(artifact string, command string) (string, error) {
+
+	err := os.Chdir(baseDir)
+	if err != nil {
+		log.Printf("reset directory back to %v error:%v\n", baseDir, err)
+	}
+
+	// tmp directory to avoid threads race on downloaded artifact
+	tmpDir := uuid.NewString()
+	err = os.Mkdir(tmpDir, 0777)
```

Comment:
I think if you use `MkdirTemp`, you won't need to use uuid.

Reply:
`MkdirTemp` is cool. refactored.